### ISSUE TITLE
Add a willFloat function to detect whether a window will float in a ManageHook

### ISSUE THREAD

**Issue by andrea-berling** [xmonad/xmonad#300]
### Problem Description

The problem is this: I would like for windows to grow in the opposite directions w.r.t. the default one (e.g. from left to right in Tabbed Simplest). To do this I used the `insertPosition` from xmonad-contrib. Now, I wanted to use it only for windows that do not float, as if I do it also for windows that do float what will happen is that an older window will stay on top of the newest, and thus some floating windows will remain hidden (it will look as if no new window was drawn at all). Thus I wanted to make an exception for floating windows in my ManageHook. Sadly `isFloat` from `XMonad.Hooks.RefocusLast` won't work, since when manageHook runs, the window isn't managed yet. Thus I wrote my own `willFloat` method based on the way XMonad decides if a window is floating or not. This does work in my manageHook and does achieve what I was aiming for. Maybe this will also be helpful to others, so I will leave the code for it below so that you can see if it may be a good method to be exported by xmonad core.

### Configuration File

```haskell
willFloat::Query Bool
willFloat = ask >>= \w -> liftX $ withDisplay $ \d -> do
  sh <- io $ getWMNormalHints d w
  let isFixedSize = isJust (sh_min_size sh) && sh_min_size sh == sh_max_size sh
  isTransient <- isJust <$> io (getTransientForHint d w)
  return (isFixedSize || isTransient)
```
Example configuration:
```haskell
module Main (main) where

import XMonad
import Data.Maybe(isJust)
import XMonad.Hooks.InsertPosition(insertPosition,Focus(..),Position(..))

willFloat::Query Bool
willFloat = ask >>= \w -> liftX $ withDisplay $ \d -> do
  sh <- io $ getWMNormalHints d w
  let isFixedSize = isJust (sh_min_size sh) && sh_min_size sh == sh_max_size sh
  isTransient <- isJust <$> io (getTransientForHint d w)
  return (isFixedSize || isTransient)

main :: IO ()
main = xmonad def
        { modMask = mod4Mask -- Use Super instead of Alt
        , terminal = "urxvt"
        , manageHook = fmap not willFloat --> insertPosition Below Newer
        }
```

### Checklist

  - [x] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - [x] I tested my configuration with [xmonad-testing](https://github.com/xmonad/xmonad-testing)

**Comment by kopr12** [comment]
I have the same problem, however I'm having a very hard time with haskell, although I'm pretty fluent with C++
Anyway, your solution is pretty good, though I would add additional check and do `insertPosition Above Newer` if window is a float
So, basically if window is a float then do `Above` else `Below`

**Comment by andrea-berling** [comment]
I think `Above Newer` is the default. I don't have such a setting in my configuration and so far I haven't experienced any issues with floating windows (all of them appear on top of the newest, as expected). I'm using the `Tall`, `Mirror Tall`, `Full` and `tabbed` layouts, but maybe the one you are using (if you're using a different one) has a different default. Anyway, adding `insertPosition Above Newer` if the window is a float doesn't hurt, so it may as well be added as far as I'm concerned.

**Comment by kopr12** [comment]
I deleted `Full` and have `Tall` and `Mirror Tall`
For some reason in my case, new float windows appear behind old window using your config
You're right, default is `Above Newer`, but my idea was to change that default and make `Below Newer` the new default  and do `Above Newer` only for float windows, problem is I don't know how to code that
My understanding is that `fmap not` is checking if window is not a float and if it's not then insert a new position, that's what you wanted right? You had a problem with floating windows if `insertPosition Below Newer` was used, older window would be on top of new window, that's why you had to implement this solution, but now if I understand you correctly you don't have that problem anymore, but I'm puzzled what changed ...
So, what I wanted to try is something different, as you can see below, the only problem is I don't know how to replace `fmap not` to be something like `willFloat --> insertPosition Above Newer`

```
myLayout = tiled ||| Mirror tiled
  where
    tiled   = Tall nmaster delta ratio
    nmaster = 1      -- Default number of windows in the master pane
    ratio   = 1/2    -- Default proportion of screen occupied by master pane
    delta   = 3/100  -- Percent of screen to increment by when resizing panes

myLayoutHook = avoidStruts 
               $ smartBorders
               $ spacingRaw False (Border 4 4 6 6) True (Border 4 4 6 6) True
               myLayout

myManageHook = composeAll
    [ isFullscreen --> doFullFloat
    , insertPosition Below Newer
    , fmap not willFloat --> insertPosition Above Newer
    , className =? "Galculator"     --> doCenterFloat
    , isDialog                      --> doCenterFloat
    ]
```

manageHook = myManageHook

**Comment by kopr12** [comment]
I mean, I said it wrong, you have it working with your solution, basically doing `insertPosition Below Newer` for every window that's not a float, but for some reason that's not working in my case, really can't tell why though, so I have to try something else.

edit: by not working I mean `Below` is applied to floating windows as well

**Comment by andrea-berling** [comment]
> You're right, default is Above Newer, but my idea was to change that default and make Below Newer the new default and do Above Newer only for float windows, problem is I don't know how to code that

I think you can achieve that with the following:
```haskell
myLayout = tiled ||| Mirror tiled
  where
    tiled   = Tall nmaster delta ratio
    nmaster = 1      -- Default number of windows in the master pane
    ratio   = 1/2    -- Default proportion of screen occupied by master pane
    delta   = 3/100  -- Percent of screen to increment by when resizing panes

myLayoutHook = avoidStruts 
               $ smartBorders
               $ spacingRaw False (Border 4 4 6 6) True (Border 4 4 6 6) True
               myLayout

myManageHook = composeAll
    [ isFullscreen --> doFullFloat
    , insertPosition Below Newer
    , fmap not willFloat --> insertPosition Below Newer
    , willFloat --> insertPosition Above Newer
    , className =? "Galculator"     --> doCenterFloat
    , isDialog                      --> doCenterFloat
    ]
```
Then again, if `Above Newer` is the default, then the line `, fmap not willFloat --> insertPosition Above Newer`  would be redundant. See if it works for you though.

**Comment by slotThe** [comment]
On Wed, Jun 16 2021 11:11, kajzersoze wrote:
> My understanding is that `fmap not` is checking if window is not a
> float and if it's not then insert a new position

No, `willFloat` checks whether the window will be floating and `fmap
not` is simply for negating that.  In this case, `willFloat` is a `Query
Bool`, so simply writing `not willFloat` would not cut it; we need to
"map over" the query part as well, which is what `fmap` will do.

> So, what I wanted to try is something different, as you can see below,
> the only problem is I don't know how to replace `fmap not` to be
> something like `willFloat --> insertPosition Above Newer`

You can really just write `willFloat --> insertPosition Above Newer` in
that case.

**Comment by kopr12** [comment]
I tried it and it still doesn't work.
The only thing which works for me is `fmap not isDialog --> insertPosition Below Newer`
Works perfect, but I had only one problem with it, some game was putting new window behind old one and I couldn't see it, I suspect it wasn't detected as 'dialog' , but I'm not sure, it's hard to test that because I would have to delete all the settings and start again, because that happens only on first launch, anyway doesn't matter ... I guess I'll stick with `fmap not isDialog....` 

btw @slotThe  thx for the explanation, I figured that much, but in my case `fmap not willFloat ->..` has the same effect as `willfloat -> ..` so I don't know, I guess that maybe for some reason in my config that function isn't working right, I don't know what's the alternative explanation, if it's working for OP

**Comment by mikenrafter** [comment]
If I'm understanding this issue thread correctly, the desire is to change which direction the windows are added, up/down stack to down/up stack.

I faced this problem, and here's my solution (used a `manageHook`, and `xmonad-contrib`):
[my `xmonad.hs` L245](https://gitlab.com/mikenrafter/config/-/blob/master/.xmonad/xmonad.hs#L245),
[my `xmonad.hs` L326](https://gitlab.com/mikenrafter/config/-/blob/master/.xmonad/xmonad.hs#L326),
[my `xmonad.hs` L359](https://gitlab.com/mikenrafter/config/-/blob/master/.xmonad/xmonad.hs#L359),
[my `lib/Custom/Keys.hs` L348](https://gitlab.com/mikenrafter/config/-/blob/master/.xmonad/lib/Custom/Keys.hs#L348)

My solution encompasses other behavior I desired, but, it could be easily stripped to fill this simpler use-case.

**Comment by truly-irham** [comment]
Thanks for this. This has been bugging me for days now lol. Works like charm.

**Comment by slotThe** [comment]
@andrea-berling I think this would be a great addition to X.ManageHook, in case you are still interested :)

**Comment by andrea-berling** [comment]
Great. I can work on a pull request 🙂

**Comment by slotThe** [comment]
@andrea-berling friendly ping :)

**Comment by andrea-berling** [comment]
@slotThe hey, sorry, I got busy in the meantime and kind of forgot about this. Working on it right now :slightly_smiling_face:

**Comment by andrea-berling** [comment]
@slotThe just created the PR. I've been a little brief in the description of the PR and basically redirected the reader to this issue, since I've already described in decent detail what this function does, why I thought it was needed/useful, and the like. I hope that's not going to be a problem. If you'd rather I actually wrote in the description of the PR the whole story just let me know, I will do that instead :slightly_smiling_face:
